### PR TITLE
Fixes #35, more details of the job

### LINKED ISSUE #35: More infos when job delievered
Mainly those:

>  * li revenue
>  * li earned_xp
>  * li delivery_time
>  * li autopark_used
>  * li autoload_used

https://github.com/Lihis/scs-sdk/blob/5b2bf461c98731687c508ef85e2082da213e5aee/include/scs/common/scssdk_telemetry_common_gameplay_events.h#L21-L33

Especiall this is quite important
> job_market

https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L142

### PR DESCRIPTION
**Pull request by Lihis**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -115,6 +121,36 @@ Below is an example of JSON format for a job.
...
+
+**Note:** Job with a status `0` should not ever be sent to the API, if that
+happens it's a bug.
+
+#### Types
+
+Description of job types:
+
+```
+0 = Unknow
```
Comment (Subtixx) [inline README.md]: This should be Unknown

**Comment by Subtixx** [comment]
Getting an exception when starting the app now

> Exception thrown at 0x00007FF78C32C772 in ets2-job-logger.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFF8.

**Comment by Lihis** [comment]
> Getting an exception when starting the app now

Works here on Linux.. Build a debug build and run with debugger to see where it crashes?

File: README.md (hunk 1 above)
Comment (Lihis) [reply README.md]: Fixed the typo but had typo in the commit message, force pushed..

**Comment by Subtixx** [comment]
Its crashing in the deallocator of a string.
img
img

**Comment by Lihis** [comment]
It can't be `trailer_t` constructor which is causing that? Remove it and let me know.

**Comment by Subtixx** [comment]
Nope still happening with the ctor commented out

File: README.md [hunk 2]
```diff
@@ -92,6 +92,10 @@ Below is an example of JSON format for a job.
         "driven": 0.0,            // Kilometers
         "planned": 248            // Planned trip distance kilometers
     },
+    "trailer": {
+        "id": ""scs_box.moving_floor.chassis_stwx2esii""
```
Comment (Subtixx) [inline README.md]: Double quotes

**Comment by Lihis** [comment]
Hmm yeah I see `id_name_t` on the third line in the first picture.. Can that backtrace from Visual Studio be copied? If it can and it include also line numbers then put it somewhere like pastebin.com.

**Comment by Subtixx** [comment]
Commenting out `m_Job  = ` line results in:
img

**Comment by Lihis** [comment]
Stacktrace was not helpful as I thought.. Just confirms the flow of:

`MainWindow::socket_run()` -> try connect plugin -> fail (game not running I assume) -> `MainWindow::socket_on_fail()` -> clear `job_t` struct and update labels.

Got me thinking is `m_job = job_t(m_job.game);` relying to undefined behavior.

**Comment by Subtixx** [comment]
Btw it broke with the commit 92bd801a0d72d6e4e5c88b4af4a7ac8cde3f9232

**Comment by Lihis** [comment]
I haven't spotted anything, yet. Something must be corrupting the memory. I did quick inspection with `valgrind` and only spotted one memory leak. Don't know would Visual Studio's `Memory Profiler` be helpful.

Anyway inspection continues.

**Comment by Lihis** [comment]
Maybe now it won't crash?

**Comment by Subtixx** [comment]
Nope. Works only until the game connects, then the application crashes

**Comment by Lihis** [comment]
Same crash but now in [MainWindow::socket_on_message()](https://github.com/Lihis/ets2-job-logger/pull/37/files#diff-656ac97937601a791c9144eaa16bbeb0R249)?

**Comment by Lihis** [comment]
@donatocl does this crash for you too?

**Comment by donatocl** [comment]
@Lihis im in the way to home, in like 20 minutes I can test and let you know 👍🏻

**Comment by donatocl** [comment]
@Lihis its working like a charm form me, no errors in ETS2 and ATS

**Comment by donatocl** [comment]
I don't know if is important, but in VS2017 a werid warning say: "gui\logger\jobsender.cpp(240): warning C4138: '*/' found outside of comment" but the APP is working without errors in ETS2 and ATS.

**Comment by Lihis** [comment]
@donatocl thanks! I wonder what is causing your issue @Subtixx...

> I don't know if is important, but in VS2017 a weird warning say: "gui\logger\jobsender.cpp(240): warning C4138: '*/' found outside of comment" but the APP is working without errors in ETS2 and ATS.

I noticed that in the CI build logs the other day but it's just warning due my coding style (didn't know it could trigger a warning but of course it's M$ compiler specific). Need to check how to get rid off it nicely..

**Comment by donatocl** [comment]
JSON post received
json_post_from_app

**Comment by donatocl** [comment]
Because is more of the same and to avoid have different issues for the same purpose.

If is possible to add information about truck to JSON post like this:

"truck": {
    "id": "[@li id](https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L67)",
    "name": "[@li name](https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L68)",
    "wheel_count": "[@li wheel_count](https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L88)",
    "brand_name": "[@li brand](https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L66)",
    "brand_id": "[@li brand_id](https://github.com/Lihis/scs-sdk/blob/922fb3a7ac271d8ec079654bf2449fb2abf103b7/include/scs/common/scssdk_telemetry_common_configs.h#L65)"
}

**Comment by Subtixx** [comment]
> Same crash but now in [MainWindow::socket_on_message()](https://github.com/Lihis/ets2-job-logger/pull/37/files#diff-656ac97937601a791c9144eaa16bbeb0R249)?

Yes
img

**Comment by Lihis** [comment]
Tested quickly that everything works as with changes for #38 came "major" changes in the plugin.

> Yes
Hmm.. Works fine on Linux and with Wine (of course Wine != Windows) and I haven't spotted anything wrong anywhere. As it works for donatocl it must be related to your environment or build configuration.

**Comment by donatocl** [comment]
Testing and there is a problem, when I cancel a job, the json post don't send info about source/destination city and company, here two json examples:

-  Starting JOB: https://pastebin.com/9hRSLgFg
- Ending JOB: https://pastebin.com/jK2f0Dxw

The JOB was taken in freight market and cancelled one minute later.

**Comment by donatocl** [comment]
Now is working!

**Comment by Lihis** [comment]
Thanks for testing!

@Subtixx your crash issue tho, that probably needs help from someone who is also using or is familiar with Windows.

**Comment by donatocl** [comment]
@Lihis I have been testing and when you cancel a job, it is canceled but immediately with the same information it is like another one starts. I reproduced it in the following way: take a job with your own trailer (cargo market) load the trailer, move a few km and cancel the job, canceling it is canceled, but seconds later it starts again. For now is just happens on cargo market and just when cancel a job.

**Comment by donatocl** [comment]
> 
> 
> @Lihis I have been testing and when you cancel a job, it is canceled but immediately with the same information it is like another one starts. I reproduced it in the following way: take a job with your own trailer (cargo market) load the trailer, move a few km and cancel the job, canceling it is canceled, but seconds later it starts again. For now is just happens on cargo market and just when cancel a job.

Just to clarify, when the job is cancelled and seconds later it starts again, is just sended with JSON post, in the gui it like no job is on going.

**Comment by Lihis** [comment]
How about we split Truck endpoint into `truck/info` and `truck/coords`?

From what I have been testing, the job would be sent first and then the truck info but the truck info would not be sent if the truck between the jobs does not change.

**Comment by donatocl** [comment]
The problem with this is how to associate that truck info with the job.

**Comment by Lihis** [comment]
My bad, it has pretty much same issue as in the plugin..

**Comment by donatocl** [comment]
This is something that can be fixed or you really need to send truck info to other endpoint? I ask just to understand, sorry to ask

**Comment by Lihis** [comment]
I don't necessarily need it, I just quickly thought the issue what I'm having could be easier solved on that end but basically it would just move the issue from the plugin to the server.

The issue is that if you change the truck then the truck in the JSON is wrong.

For example:
1. You have e.g. Scania Streamline as your truck
2. Take a quick job where truck is e.g. Volvo
3. Cancel the job

Expected result is the truck is Volvo in the JSON.
Actual result is that the truck is Scania in the JSON.

This happens because when you click "Yes" to cancel the job the plugin will:

1. Receive new configuration
    - Update truck
2. Receive cancel event
    - Send the job to the GUI

**Comment by donatocl** [comment]
> 
> 
> I don't necessarily need it, I just quickly thought the issue what I'm having could be easier solved on that end but basically it would just move the issue from the plugin to the server.
> 
> The issue is that if you change the truck then the truck in the JSON is wrong.
> 
> For example:
> 
>     1. You have e.g. Scania Streamline as your truck
> 
>     2. Take a quick job where truck is e.g. Volvo
> 
>     3. Cancel the job
> 
> 
> Expected result is the truck is Volvo in the JSON.
> Actual result is that the truck is Scania in the JSON.
> 
> This happens because when you click "Yes" to cancel the job the plugin will:
> 
>     1. Receive new configuration
>        
>        * Update truck
> 
>     2. Receive cancel event
>        
>        * Send the job to the GUI

Sorry for the delay in answer. Thank you a lot  @Lihis for explain with more details!

So basically this is also causing the problem described in my comments right? (resend json post with status 1 after cancel a cargo market job)

> The issue is that if you change the truck then the truck in the JSON is wrong.
So, this happen just when cancel a job or also when finish a job?

**Comment by Lihis** [comment]
> So basically this is also causing the problem described in my comments right? (resend json post with status 1 after cancel a cargo market job)

No this is not causing that. While I was fixing that and testing that the changes work correctly I noticed this issue. I'll get back to this tomorrow.

**Comment by Lihis** [comment]
Copy-paste from what I added now to [README.md](https://github.com/Lihis/ets2-job-logger/pull/37/commits/a8328f60bc132139bbf6443c4f6f0723866de408):

> - Truck _can be_ wrong when status is `Delivered` or `Cancelled`
>     - e.g. when user owns a truck but decides to do a quick job; truck is
>       correct when `OnJob` but when `Delivered`/`Cancelled` the truck is
>       reported as the truck the user owns instead of the one used in the job.

It seems that you can't go to a truck dealer and trade in your current truck so in that sense the truck stays static from beginning of the job to the end of the job. Editing truck configuration is possible and by replacing chassis the amount of wheel could change but I don't know if the `id` is related to the cabin or something so could it change during the job. Truck configuration however can't be edited when doing quick job as "employer might not like tinkering with the configuration".

Knowing that the amount of wheels changed _during_ the job probably is not interesting information?

---

I'm thinking if the Job API should be refactored as in my opinion currently it contains plenty of stuff what is only relevant with specific status. Something like `job/new`, `job/cancelled` and `job/delivered` would make more sense and be more REST like. It would result a nicer implementations as code could be then split in the PHP frameworks routing level?

**Comment by donatocl** [comment]
> I'm thinking if the Job API should be refactored as in my opinion currently it contains plenty of stuff what is only relevant with specific status. Something like job/new, job/cancelled and job/delivered would make more sense and be more REST like. It would result a nicer implementations as code could be then split in the PHP frameworks routing level?
This will be really a good idea! But I know that will be lot of work for you, sorry for not know about language to help you :(

> Knowing that the amount of wheels changed during the job probably is not interesting information?
Nope, and if you need to remove the wheels, do it, not a problem at all.

**Comment by Lihis** [comment]
> This will be really a good idea!

Great :) then we will merge this PR and I'll create an issue of the refactoring (this PR is big already and doing refactoring also in this one would make this huge).

> Nope, and if you need to remove the wheels, do it, not a problem at all.

It would be nice to have; keep it but the value is what it is when starting the job.

**Comment by donatocl** [comment]
> 
> 
> > This will be really a good idea!
> 
> Great :) then we will merge this PR and I'll create an issue of the refactoring (this PR is big already and doing refactoring also in this one would make this huge).
> 
> > Nope, and if you need to remove the wheels, do it, not a problem at all.
> 
> It would be nice to have; keep it but the value is what it is when starting the job.

Thank you!

**Comment by Lihis** [comment]
@Subtixx if you still have that problem and clean build (i.e. manually remove the build directory) doesn't help then create an issue.

**Pull request merged**